hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -1957,17 +1959,27 @@ public boolean normalizeRegions() throws IOException {
             continue;
           }
 
-          // as of this writing, `plan.execute()` is non-blocking, so there's no artificial rate-
-          // limiting of merge requests due to this serial loop.
+          // as of this writing, `plan.submit()` is non-blocking and uses Async Admin APIs to
+          // submit task , so there's no artificial rate-
+          // limiting of merge/split requests due to this serial loop.
           for (NormalizationPlan plan : plans) {
-            plan.execute(admin);
+            Future<Void> future = plan.submit(admin);
+            submittedPlanList.add(future);
             if (plan.getType() == PlanType.SPLIT) {
               splitPlanCount++;
             } else if (plan.getType() == PlanType.MERGE) {
               mergePlanCount++;
             }
           }
         }
+        for (Future<?> submittedPlan : submittedPlanList) {
+          try {
+            submittedPlan.get();
```

Comment:
This changes behaviors for normalizer. Can we give some timeout for blocking so it wont block forever if something goes wrong with one of plans?

Reply:
True, there are either of these options we can follow:
1. Timeout with Future.get()
2. Perform all .get() operations asynchronously

With huge no of submitted plans, do you think it makes sense to even perform `.get()` asynchronously by ThreadPoolExecutor and let it block until all plans are done and even update `normalizationPlanFailureCount` count. What do you think?

Reply:
I don't think the counter as a metric will help an operator very much. Regions can split or merge in the background, causing a planed action to fail. The failure should be logged, I think (maybe the normal split/merge pathways do this already? I haven't checked).

It would be nice to have a final log message for the normalizer run that says something like "successfully executed x of y normalization actions."

Reply:
As Nick commented, need to think about the purpose of this change. I think the purpose is to know the result of plans, which comes with cost.
1). Timeout with Future.get().
    When it times out, we do not know if plan succeeds or not. The only info it gives us is that the plan does not finish within a certain amount of time.
2). Perform all.get() asynchronously.
    If get() blocks for whatever reason, there will be huge number of threads blocking for get(), system resource leak.

   Maybe the best approach is per Nick's comments, log how many plans submitted. If some plans fail, we can go to procedure system for root cause.